util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,249 @@
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(
+            "Failed to obtain resources via url request. "
+            f"Exception thrown:\n{e}"
+        )
+
+    json_response = json.loads(response.read().decode("utf-8"))
+    assert "documents" in json_response, (
+        "'documents' not a key in "
+        "json dictionary. JSON: \n"
+        f"{json_response}"
+    )
+    resources = json_response["documents"]
+
+    return resources
+
+
+def save_resources_to_file(resources: List[Dict], output: TextIO):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def get_resource_local_filepath(
+    resource_id: str, resource_version: str, base_output_path: TextIO
+) -> Path:
+    """
+    This function returns the local filepath for a resource
+    :param resource_url: URL field of the resource
+    :param base_output_path: Base output directory absolute path
+    :return: Local filepath for the resource
+    """
+    filename = f"{resource_id}-{resource_version}"
+
+    filepath = Path(base_output_path).joinpath(filename)
+
+    return filepath
+
+
+def download_resources(resources: List[Dict], output: TextIO):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filepath = get_resource_local_filepath(
+                resource["id"], resource["resource_version"], path
+            )
+
+            if not filepath.exists() or (  # If the file does not exist
+                "url-md5sum"
+                in resource.keys()  # If it exist but the md5sum is wrong.
+                and hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["url-md5sum"]
+            ):
+                print(f"Downloading {url} to {filepath}")
+                request.urlretrieve(url, filepath)
+            resource["url"] = filepath.absolute().as_uri()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
```

Comment:
This comment applies across a lot of the documentation/comments in this PR, but I'll bring it up here as this is a good example:

We are overloading the  "resource" and I fear it's going to confuse people. I've mentioned it before but we use it to mean 3 different parts of the infrastructure:

1. The entry in the database/JSON that we reference to find the disk images, binaries, etc. we need. (Small note here: from my understanding MongoDB uses the term "documents" to refer to what I think of as "entries").
2. The file that's obtained from the URL in the database entry (two small caveats here: (A) some of these are compressed or archived directories and therefore need further processing to be used in a simulation, is a resource just the downloaded file or the file after the decompressing and unarchiving operations?; and (B) some resources don't even have a file that needs downloaded. All the data they need is in the database entry. `SimpointResource` works like this. What would it mean to "download a SimpointResource", for example?
3. The `object` constructed from the `Resource` classes. I.e., the object returned by the `obtain_resource` function.



I'm open to alternative viewpoints here but I think a "gem5 Resource" makes most sense being the Python object. That's what's ultimately passed to the simulation configuration. This object is constructed from the database entry/document. The resource object may reference a "remote file" (or files -- we shouldn't limit ourself here) but this is just a design quirk. A resource is still a resource whether or not it a "remote file" it needs to download for the simulator.

How about:

* The gem5 Resource object: A Python object passed to the simulation to configure what simulated hardware runs, and how it is run.
* The gem5 Resource data-source entry: The enerty in the database (in most cases gem5's MongoDB database) which defines the object and is referenced by the Standard Library to construct a gem5 Resource object.
* A gem5 resource remote file: A remote file referenced by a gem5 Resource object which may be downloaded and used in a simulation. For example, the `x86-ubuntu-22.04-img` resource is dependent on the a remote disk image containing X86 Ubuntu 22.04 but this alone is not the "resource" as information on the Root partition is required by the simulator. It is the gem5 Resources object which encapsulates everything which is needed.

Reply:
I agree with all of this. I'll just add that, colloquially, when I say "we need a gem5 resource" or "get the resource" I am referring to the artifact that is used with gem5. E.g., a disk image is a resource. 

However, the documentation should be more precise, and I like how Bobby describes it here.